tests/integration-tests/tests/integration/catalog_tests.rs

```diff
@@ -560,6 +561,134 @@ async fn test_axfr_refused() {
+    // The catalog had no payload configured.
+    assert!(
+        edns.option(EdnsCode::NSID).is_none(),
+        "unexpected NSID in reply EDNS"
+    );
+}
+
+// Test that **not** requesting NSID produces no NSID response, even when a payload is configured.
+#[tokio::test]
+async fn test_nsid_enabled_not_requested() {
+    subscribe();
+
+    let mem_authority = create_test();
+    let origin = mem_authority.origin().clone();
+    let mut catalog = Catalog::new();
+    catalog.upsert(origin.clone(), vec![Arc::new(mem_authority)]);
+
+    // Configure the catalog with an NSID payload.
+    catalog.set_nsid(Some(NSIDPayload::new(vec![0xC0, 0xFF, 0xEE]).unwrap()));
+
+    // Create a question request that doesn't ask for NSID in EDNS.
+    let question_req = test_nsid_request(origin.clone(), false);
+
+    let response_handler = TestResponseHandler::new();
+    let _ = catalog
+        .handle_request(&question_req, response_handler.clone())
+        .await;
+    let response = response_handler.into_message().await;
+    assert_eq!(response.response_code(), ResponseCode::NoError);
+
+    // We sent EDNS in the request, and so expect to find EDNS in the response.
+    let edns = response
+        .extensions()
+        .as_ref()
+        .expect("missing response EDNS");
+    // We shouldn't find an NSID payload in the response EDNS - we didn't request it.
+    assert!(
+        edns.option(EdnsCode::NSID).is_none(),
+        "unexpected NSID in reply EDNS"
+    );
+}
+
+// Test that requesting NSID when a payload is configured produces the expected payload
+// in the response.
+#[tokio::test]
+async fn test_nsid_enabled_and_requested() {
+    subscribe();
+
+    let mem_authority = create_test();
+    let origin = mem_authority.origin().clone();
+    let mut catalog = Catalog::new();
+    catalog.upsert(origin.clone(), vec![Arc::new(mem_authority)]);
+
+    // Configure the catalog with an NSID payload.
+    let nsid = NSIDPayload::new(vec![0xC0, 0xFF, 0xEE]).unwrap();
+    catalog.set_nsid(Some(nsid.clone()));
+
+    // Create a question request that asks for NSID in EDNS.
+    let question_req = test_nsid_request(origin.clone(), true);
+
+    let response_handler = TestResponseHandler::new();
+    let _ = catalog
+        .handle_request(&question_req, response_handler.clone())
+        .await;
+    let response = response_handler.into_message().await;
+    assert_eq!(response.response_code(), ResponseCode::NoError);
+
+    // We sent EDNS in the request, and so expect to find EDNS in the response.
+    let edns = response
+        .extensions()
+        .as_ref()
+        .expect("missing response EDNS");
+    // We should find the expected EDNS NSID payload.
+    assert_eq!(edns.option(EdnsCode::NSID), Some(&EdnsOption::NSID(nsid)));
+}
+
+fn test_nsid_request(origin: LowerName, request_nsid: bool) -> Request {
+    let mut query: Query = Query::new();
```

Comment:
Nit: guessing you copy/pasted this from somewhere, please kill the type annotation.

Reply:
> guessing you copy/pasted this from somewhere

Guilty. It was this way throughout most of the pre-existing tests. I added a commit to clean up the other instances and then fixed my new helper function to avoid it as well. Thanks!